### PR TITLE
Playwright 🎭  🛠️ 

### LINKED ISSUE #30761: Test Failure: BlazorServerTemplateWorks_NoAuth
Build: https://dev.azure.com/dnceng/public/_build/results?buildId=1027022&view=logs&j=8c9dd6ae-5017-5a70-997c-c2386fed4e85&t=67839355-6a03-55d6-fbd7-5d56c2fd9ea1

```
  Failed Templates.Test.BlazorServerTemplateTest.BlazorServerTemplateWorks_NoAuth(browserKind: Chromium) [2 m 50 s]
  Error Message:
   System.TimeoutException : Timeout 30000ms exceeded.
=========================== logs ===========================
waiting for selector "h1+p >> text=Current count: 1" to be visible
============================================================
Note: use DEBUG=pw:api environment variable and rerun to capture Playwright logs.
  Stack Trace:
     at PlaywrightSharp.Transport.Connection.SendMessageToServerAsync[T](String guid, String method, Object args, Boolean ignoreNullValues, JsonSerializerOptions serializerOptions, Boolean treatErrorPropertyAsError) in /Users/neo/Documents/Coding/Personal/playwright-sharp/src/PlaywrightSharp/Transport/Connection.cs:line 163
   at PlaywrightSharp.Frame.WaitForSelectorAsync(Boolean isPageCall, String selector, Nullable`1 state, Nullable`1 timeout) in /Users/neo/Documents/Coding/Personal/playwright-sharp/src/PlaywrightSharp/Frame.cs:line 645
   at Templates.Test.BlazorServerTemplateTest.TestBasicNavigation(IPage page) in /_/src/ProjectTemplates/BlazorTemplates.Tests/BlazorServerTemplateTest.cs:line 200
   at Templates.Test.BlazorServerTemplateTest.BlazorServerTemplateWorks_NoAuth(BrowserKind browserKind) in /_/src/ProjectTemplates/BlazorTemplates.Tests/BlazorServerTemplateTest.cs:line 74
   at Templates.Test.BlazorServerTemplateTest.BlazorServerTemplateWorks_NoAuth(BrowserKind browserKind) in /_/src/ProjectTemplates/BlazorTemplates.Tests/BlazorServerTemplateTest.cs:line 101
```

### LINKED ISSUE #30882: Flaky Various Blazor Template timeouts/failures
It looks like almost any blazor template test can fail due to a shared fixture being used between them and it timing out.

Builds: https://dev.azure.com/dnceng/internal/_build/results?buildId=1033436&view=ms.vss-test-web.build-test-results-tab&runId=32069600&resultId=125477&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

https://dev.azure.com/dnceng/public/_build/results?buildId=1035875&view=ms.vss-test-web.build-test-results-tab&runId=32132094&resultId=125303&paneView=debug

```
System.TimeoutException : Timeout 30000ms exceeded.
=========================== logs ===========================
<launching> C:\h\w\AC6D092A\w\A51C0945\e\ms-playwright\firefox-1225\firefox\firefox.exe -no-remote -headless -profile C:\h\w\AC6D092A\t\playwright_firefoxdev_profile-0AKzH5 -juggler-pipe -silent
<launched> pid=4752
[err] *** You are running in headless mode.
[out]
[out] Juggler listening to the pipe
[out] console.warn: SearchSettings: "get: No settings file exists, new profile?" (new Error("", "(unknown module)"))
[err] JavaScript error: resource://gre/modules/XULStore.jsm, line 66: Error: Can't find profile directory.
[err]
[err] ###!!! [Child][RunMessage] Error: Channel closing: too late to send/recv, messages will be lost
[err]
Note: use DEBUG=pw:api environment variable and rerun to capture Playwright logs.
```

### PR DESCRIPTION
**Pull request by TanayParikh**
Enable Playwright Tests

Fixes: https://github.com/dotnet/aspnetcore/issues/30761
Fixes: https://github.com/dotnet/aspnetcore/issues/30882

### CODE REVIEW

File: src/ProjectTemplates/README.md
Comment (mkArtakMSFT) [inline src/ProjectTemplates/README.md:74]: Based on the title it seems like this is for running all playwright tests, but it's referring to only the `Blazor Template Tests`. Should the title be updated?

File: src/Shared/Process/ProcessEx.cs
Comment (TanayParikh) [inline src/Shared/Process/ProcessEx.cs:182]: We're already in the `Exited` Event so this call is redundant:

> There are two ways of being notified when the associated process exits: synchronously and asynchronously. Synchronous notification means calling the WaitForExit method to block the current thread until the process exits. Asynchronous notification uses the Exited event, which allows the calling thread to continue execution in the meantime. In the latter case, EnableRaisingEvents must be set to true for the calling application to receive the Exited event.

https://learn.microsoft.com/dotnet/api/system.diagnostics.process.exited?view=net-7.0

File: src/ProjectTemplates/test/Templates.Blazor.Tests/BlazorWasmTemplateTest.cs
Comment (TanayParikh) [inline src/ProjectTemplates/test/Templates.Blazor.Tests/BlazorWasmTemplateTest.cs:62]: Will replace with issue & tackle separately.

**Review by MackinnonBuck — APPROVED** [review]

File: .azure/pipelines/ci.yml
Comment (mkArtakMSFT) [inline .azure/pipelines/ci.yml:700]: Will setting this new variable (RunBlazorPlaywrightTemplateTests) to `false` result in the current behavior (as if we don't have this change in place)?

File: src/ProjectTemplates/test/Templates.Blazor.Tests/BlazorWasmTemplateTest.cs
Comment (mkArtakMSFT) [inline src/ProjectTemplates/test/Templates.Blazor.Tests/BlazorWasmTemplateTest.cs:110]: The issue here was blocked on a Playwright issue, which they have resolved with the following comment: 
image

Does that impact our decisions regarding this in any way?
I assume you're trying to wrap up your work already, and leaving this to be handled as part of the work on the linked issue is the right thing to do. So mostly wanted to note the progress on the dependency.

**Review by mkArtakMSFT — APPROVED** [review]

File: .azure/pipelines/ci.yml
Comment (TanayParikh) [reply .azure/pipelines/ci.yml:700]: It's fine to just omit `RunBlazorPlaywrightTemplateTests` in that case as it'll default to using the value of `RunTemplateTests`, but yes setting it to `false` would work as well.

https://github.com/dotnet/aspnetcore/blob/20ebf882bf561130d3288a419ee2caa6f3d343c4/src/ProjectTemplates/test/Templates.Blazor.Tests/Templates.Blazor.Tests.csproj#L10-L11

File: src/ProjectTemplates/test/Templates.Blazor.Tests/BlazorWasmTemplateTest.cs
Comment (TanayParikh) [reply src/ProjectTemplates/test/Templates.Blazor.Tests/BlazorWasmTemplateTest.cs]: CI Issues tracked in https://github.com/dotnet/aspnetcore/issues/46430

**Comment by wtgodbe** [comment]
The problem is that the `Microsoft.AspNetCore.Components.WebAssembly` package hasn't been built yet at the time the tests run (other packages like `Microsoft.AspNetCore.Components.WebAssembly.Server` _are_ available). Trying to figure out if that always would have been the case or if something's changed here.

**Review by wtgodbe — APPROVED** [review]
LGTM

File: src/ProjectTemplates/test/Templates.Blazor.Tests/BlazorWasmTemplateTest.cs
Comment (wtgodbe) [reply src/ProjectTemplates/test/Templates.Blazor.Tests/BlazorWasmTemplateTest.cs]: Going to try un-skipping this

Comment (wtgodbe) [reply src/ProjectTemplates/test/Templates.Blazor.Tests/BlazorWasmTemplateTest.cs:110]: Nope, still failing

File: src/Shared/Process/ProcessEx.cs
Comment (BrennanConroy) [reply src/Shared/Process/ProcessEx.cs:182]: Pretty sure this is still needed when using async IO. It guarantees that stdout and stderr are done.
Data point; after this PR was merged there are now failures like:
```
2023-02-14T17:40:24.0586118Z [xUnit.net 00:00:04.23]         | [3.247s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: Process exited.
2023-02-14T17:40:24.0586151Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR] No templates or subcommands found matching: 'reactredux'.
2023-02-14T17:40:24.0586182Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR] 
2023-02-14T17:40:24.0586210Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR] To list installed templates similar to 'reactredux', run:
2023-02-14T17:40:24.0594633Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR]    dotnet new list reactredux
2023-02-14T17:40:24.0627195Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR] To search for the templates on NuGet.org, run:
2023-02-14T17:40:24.0629366Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR]    dotnet new search reactredux
2023-02-14T17:40:24.0736529Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR] 
2023-02-14T17:40:24.0754965Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR] 
2023-02-14T17:40:24.0774833Z [xUnit.net 00:00:04.23]         | [3.248s] Templates.Blazor.Test.BlazorWasmTemplateTest Information: [ERROR] For details on the exit code, refer to https://aka.ms/templating-exit-codes#103
2023-02-14T17:40:24.0794693Z [xUnit.net 00:00:04.23]         | [3.254s] Templates.Blazor.Test.BlazorWasmTemplateTest Error: Test threw an exception.
2023-02-14T17:40:24.0812567Z [xUnit.net 00:00:04.23]         | System.InvalidOperationException: Failed to uninstall previous templates. The template 'reactredux' could still be found.
2023-02-14T17:40:24.0901861Z [xUnit.net 00:00:04.23]         |    at Templates.Test.Helpers.TemplatePackageInstaller.VerifyCannotFindTemplateAsync(ITestOutputHelper output, String templateName) in /_/src/ProjectTemplates/Shared/TemplatePackageInstaller.cs:line 151
2023-02-14T17:40:24.0920908Z [xUnit.net 00:00:04.23]         |    at Templates.Test.Helpers.TemplatePackageInstaller.InstallTemplatePackages(ITestOutputHelper output) in /_/src/ProjectTemplates/Shared/TemplatePackageInstaller.cs:line 114
```
You can see from the test output that the text "No templates or subcommands found matching:" was output, however the test threw on the check for that exact string:
https://github.com/dotnet/aspnetcore/blob/fb1e6a74373f1e4f09cc1b0adaac5f0535e5f07d/src/ProjectTemplates/Shared/TemplatePackageInstaller.cs#L149-L151
This implies a race condition where the process exits and we grab the output before the stdout/stderr completed writing.
cc @wtgodbe

Comment (wtgodbe) [reply src/Shared/Process/ProcessEx.cs:182]: Thanks for catching this, I'll put up a PR to revert this line

**Pull request merged**
